File: Python/Medium/MinStack.py
class MinStack(object):

    def __init__(self):
        self.minValArray = []
        self.stackArray = []

    def push(self, val):
        """
        :type val: int
        :rtype: None
        """
        self.stackArray.append(val)
        self.minValArray.append(min(val, self.minValArray[-1] if self.minValArray else val))
        
    def pop(self):
        """
        :rtype: None
        """
        self.stackArray.pop()
        self.minValArray.pop()

    def top(self):
        """
        :rtype: int
        """
        return self.stackArray[-1]

    def getMin(self):
        """
        :rtype: int
        """
        return self.minValArray[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: Medium/MinStack.cpp
#include <vector>

using namespace std;

/**
 * Time Complexity of all functions are: O(1)
 * Space Complexity: O(n)
 */
class MinStack
{
private:
    vector<int> minValVec;
    vector<int> stackVec;

public:
    MinStack()
    {
    }

    void push(int val)
    {
        stackVec.push_back(val);
        minValVec.push_back(min(val, minValVec.size() ? minValVec.back() : val));
    }

    void pop()
    {
        stackVec.pop_back();
        minValVec.pop_back();
    }

    int top()
    {
        return stackVec.back();
    }

    int getMin()
    {
        return minValVec.back();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */